src/libutil/serialise.hh

```diff
@@ -283,6 +283,26 @@ struct LengthSink : Sink
     }
 };
 
+/**
+ * A wrapper source that counts the number of bytes read from it.
+ */
+struct LengthSource : Source
+{
+    Source & next;
+
+    LengthSource(Source & next) : next(next)
+    { }
+
+    uint64_t total = 0;
+
+    size_t read(char * data, size_t len) override
+    {
+        auto n = next.read(data, len);
+        total += n;
+        return n;
+    }
+};
```

Comment:
thoughts, optional

Previously we've used such a source or sink that warned as soon as the threshold was passed. I believe it counted the wrong thing, but that doesn't seem to be the case with its caller in this PR.

The benefit of warning as we go is that we can catch mistakes before the user has to wait for their store to fill up.
(For example, the backport action will instruct users to create `nixpkgs/.worktree`, which blows up `${nixpkgs}` massively if not fetched via git, and it takes a long time to copy and produces no diagnostic feedback during that time)

Possible scheme and defaults:
- Add the first 1GB silently
- Print when passing 1GB
- Print size when larger than 256 MB  (nixpkgs seems to be around 170)

Example log (brief)
```
copying very large path /home/user/nixpkgs to the store. Is this what you intended? See `<url>#conf-large-path-warning-limit`
copied /home/user/nixpkgs to the store (2.3 GiB)
```

Reply:
Another helpful feature is to have an enforced limit, which will also let us print a backtrace or launch the debugger.

Reply:
> The benefit of warning as we go is that we can catch mistakes before the user has to wait for their store to fill up.

Nowadays this is already somewhat handled by the progress bar (which will show "copying <bla> to the store" for a long time if it's a large path). It could be improved by showing the amount of data copied, like a download.